feat(client) ts-migrate LearnLayout Component

sharmaabhinav opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [ ] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #XXXXX

<!-- Feel free to add any additional description of changes below this line -->

sharmaabhinav committed:
feat(client) Renamed Learn.js to learn.tsx

sharmaabhinav committed:
feat(client) ts-migrate LearnLayout Component

thecodingaviator commented on client/src/components/layouts/learn.tsx:
```diff
@@ -35,7 +44,17 @@ const mapDispatchToProps = {
 
-class LearnLayout extends Component {
+type LearnLayoutProps = {
+  isSignedIn: boolean;
+  fetchState: fetchState;
+  user: user;
+  tryToShowDonationModal: () => void;
+  children: React.ReactNode;
```
Try to leave optional props as optional

thecodingaviator replied:
Try to leave optional props as optional

thecodingaviator reviewed (commented):
The optional props issue looks good now. Fix the lining and this should be good to go

ShaunSHamilton commented on client/src/components/layouts/learn.tsx:
```diff
@@ -18,11 +17,21 @@ import './prism-night.css';
 import 'react-reflex/styles.css';
 import './learn.css';
 
+type fetchState = {
```
```suggestion
type FetchState = {
```

ShaunSHamilton commented on client/src/components/layouts/learn.tsx:
```diff
@@ -18,11 +17,21 @@ import './prism-night.css';
+  errored: boolean;
+};
+
+type user = {
```
```suggestion
type User = {
```

ShaunSHamilton commented on client/src/components/layouts/learn.tsx:
```diff
@@ -18,11 +17,21 @@ import './prism-night.css';
   isSignedInSelector,
   userSelector,
-  (fetchState, isSignedIn, user) => ({
+  (fetchState: fetchState, isSignedIn, user: user) => ({
```
```suggestion
  (fetchState: FetchState, isSignedIn, user: User) => ({
```

ShaunSHamilton commented on client/src/components/layouts/learn.tsx:
```diff
@@ -35,7 +44,17 @@ const mapDispatchToProps = {
-class LearnLayout extends Component {
+type LearnLayoutProps = {
+  isSignedIn?: boolean;
+  fetchState: fetchState;
+  user: user;
```
```suggestion
  fetchState: FetchState;
  user: User;
```

ShaunSHamilton reviewed (changes requested):
Just nitpicks for consistency

ShaunSHamilton committed:
rename types for convention

sharmaabhinav committed:
fix(client) disable ts error for DonationModal usage

- DonationModal migration required typescript to fix error hence disable ts error in LearnLayout component

sharmaabhinav commented on client/src/components/layouts/learn.tsx:
```diff
@@ -88,6 +88,8 @@ class LearnLayout extends Component<LearnLayoutProps> {
         </Helmet>
         <main id='learn-app-wrapper'>{children}</main>
+        {/* eslint-disable-next-line @typescript-eslint/ban-ts-comment */
+        /* @ts-ignore  */}
```
Migration of `DonationModal.tsx` is required to fix hence added this a last option as per guidelines

ShaunSHamilton reviewed (approved):
This LGTM 👍

ShaunSHamilton commented:
@thecodingaviator Feel free to merge, if you are happy 🙂

thecodingaviator reviewed (commented):
LGTM!
